File: Arrays/Solutions/Product_of_Array_Except_Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)

        # output[i] will store product of all elements to the LEFT of i
        output = [0] * n

        # For index 0, there are no elements to the left
        output[0] = 1

        # First pass: build prefix products
        for i in range(1, n):
            output[i] = output[i - 1] * nums[i - 1]

        # right will store product of all elements to the RIGHT of i
        right = 1

        # Second pass (from right to left): multiply suffix products
        for i in range(n - 1, -1, -1):
            output[i] *= right
            right *= nums[i]

        return output

File: Arrays/Solutions/Product_of_Array_Except_Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();

        // output[i] will store product of all elements to the LEFT of i
        vector<int> output(n, 1);

        // First pass: compute prefix products
        for (int i = 1; i < n; i++) {
            output[i] = output[i - 1] * nums[i - 1];
        }

        // right stores product of all elements to the RIGHT of i
        int right = 1;

        // Second pass: multiply suffix products
        for (int i = n - 1; i >= 0; i--) {
            output[i] *= right;
            right *= nums[i];
        }

        return output;
    }
};
